plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/Parser.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * DBeaver - Universal Database Manager
+ * Copyright (C) 2010-2022 DBeaver Corp and others
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jkiss.dbeaver.parser;
+
+import java.util.*;
+
+import org.jkiss.dbeaver.parser.grammar.GrammarInfo;
+import org.jkiss.dbeaver.parser.grammar.nfa.GrammarNfaOperation;
+
+public class Parser {
+    private final GrammarInfo grammar;
+    private final ParserFsm fsm;
+
+    public Parser(GrammarInfo grammar, ParserFsm fsm) {
+        this.grammar = grammar;
+        this.fsm = fsm;
+    }
+
+    /**
+     * Parse text
+     * @param text
+     * @return result of parsing represented with discovered valid sequences of terminals
+     */
+    public ParseResult parse(String text) {
+        Deque<ParserState> queue = new ArrayDeque<>();
+        for (ParserFsmNode initialState : fsm.getInitialStates()) {
```

Comment:
If `fsm.getInitialStates()` returns `Collection` (which is easy to do), we might preallocate memory for the Deque

Reply:
Amount of initial states has nothing common with amount of active states during parsing. It is ok with default preallocation.